### PR TITLE
Add support for rpm packages

### PR DESCRIPTION
**Pull request by Dark-Passenger**
This is mostly copy paste job of debian packaging with minor changes for redhat systems. Most of the heavy lifting is being done by electron-installer-redhat.

### CODE REVIEW

**Comment by Dark-Passenger** [comment]
Fixes issue #1424

File: scripts/build/electron-installer-redhat-linux.sh [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+set -u
+set -e
+
+OS=$(uname)
+if [[ "$OS" != "Linux" ]]; then
+  echo "This script is only meant to be run in GNU/Linux" 1>&2
+  exit 1
+fi
+
+ELECTRON_INSTALLER_REDHAT="./node_modules/.bin/electron-installer-redhat"
```
Comment (lurch) [inline scripts/build/electron-installer-redhat-linux.sh]: For this to work as intended, you'll also need to add `electron-installer-redhat` to `package.json`.

File: scripts/build/electron-installer-redhat-linux.sh [hunk 2]
```diff
@@ -0,0 +1,75 @@
+#!/bin/bash
+
+###
+# Copyright 2016 resin.io
```
Comment (lurch) [inline scripts/build/electron-installer-redhat-linux.sh]: The year should be updated for new files

**Comment by lurch** [comment]
Thanks! I'll have to install Fedora into a VM and do some testing of this :-)

@jviotti Once this has been tested and merged, we should also host the RPM packages on bintray.

**Comment by jviotti** [comment]
Wow, this is awesome. Thanks a lot @Dark-Passenger! I'll be giving this a go soon (I have to get Fedora as well :P)

File: scripts/build/electron-installer-redhat-linux.sh (hunk 1 above)
Comment (Dark-Passenger) [inline scripts/build/electron-installer-redhat-linux.sh]: Done.

Comment (lurch) [inline scripts/build/electron-installer-redhat-linux.sh]: Whoops, my mistake, that now causes CI errors in `npm-shrinkwrap.json`.
Can you rebase this PR off the latest version of `master` (I'm afraid there's been quite a few changes to the Etcher codebase in the build-up to the 1.0.0 release), and then move `electron-installer-redhat` from `package.json` to the Dockerfile templates, as was done for `electron-installer-debian` in #1379 ?

If that gets too complicated, then just let me know and I can make those changes for you.

**Comment by Dark-Passenger** [comment]
@lurch I think i have done the dependency patches correctly. So should be good.

File: scripts/build/electron-installer-redhat-linux.sh (hunk 1 above)
Comment (Dark-Passenger) [inline scripts/build/electron-installer-redhat-linux.sh]: I think I did all the changes required. The build is passing so maybe its all done ?

**Comment by lurch** [comment]
According to https://github.com/unindented/electron-installer-redhat it depends on `rpmbuiild`. Could you add a dependency-check for that command to `scripts/build/electron-installer-redhat-linux.sh` and then add `rpm` to the list of packages installed by `apt-get` in the Dockerfiles?

If you don't mind, I won't merge this PR until after #1365 has been merged, as that'll ensure that the installer-creation also gets tested by the CI servers :-)

**Comment by lurch** [comment]
@jviotti What's going on with the CI tests here? Either we're missing a Travis branch build, or we shouldn't be expecting an Appveyor branch build, since the original branch is in @Dark-Passenger 's repo and not in this repo?

@hedss With all the recent work you've been doing with Github and status-checks, maybe you could give us a quick hint? :)

**Comment by hedss** [comment]
@lurch I think this may be down to the fact that the AppVeyor YAML config for etcher doesn't explicitly specify just Resin only branches. This may help: https://www.appveyor.com/docs/branches/

**Comment by lurch** [comment]
> This may help: https://www.appveyor.com/docs/branches/

Thanks. Looks like that's just about filtering specific branch _names_ though, not the repos which they come from. And we obviously have no idea in advance which branch-names will get used in PRs!

I'll wait to see if @jviotti has any ideas when he comes back online (since he has access to settings on this repo that I don't), and if not I'll try getting in touch with Appveyor support.

**Comment by lurch** [comment]
I've now built the x86_64 RPM package using this PR, and when I copy it to a 'fresh' Fedora25 VM (VM snapshots are _really_ useful here!), the RPM installs okay (along with a huge list of dependencies) but `etcher-electron` still doesn't run, due to the missing screensaver library issue discussed in https://github.com/resin-io/etcher/issues/835
So I guess you need to add that to the RPM's dependency list? @Dark-Passenger

**Comment by Dark-Passenger** [comment]
@lurch the dependency issue should be fixed now. though I am building now so lets see.

**Comment by Dark-Passenger** [comment]
All still works for me... though I can uninstall the dependency to check... let me see if I can get my hands on a vm to try it out soon.

**Comment by Dark-Passenger** [comment]
Tested on centos and package works here too so awesome. Still need to find a fresh install to test with.

File: Makefile [hunk 3]
```diff
@@ -153,10 +153,13 @@ endif
...
 
+
+TARGET_ARCH_REDHAT = $(shell ./scripts/build/architecture-convert.sh -r $(TARGET_ARCH) -t redhat)
+
 PRODUCT_NAME = etcher
 APPLICATION_NAME_LOWERCASE = $(shell echo $(APPLICATION_NAME) | tr A-Z a-z)
 APPLICATION_VERSION_DEBIAN = $(shell echo $(APPLICATION_VERSION) | tr "-" "~")
-
+APPLICATION_VERSION_REDHAT = $(shell echo $(APPLICATION_VERSION) | tr "-" "~")
 # Fix hard link Appveyor issues
```
Comment (lurch) [inline Makefile]: Can you add the blank line back in here (above the comment) please?

File: Makefile [hunk 4]
```diff
@@ -153,10 +153,13 @@ endif
 
 TARGET_ARCH_DEBIAN = $(shell ./scripts/build/architecture-convert.sh -r $(TARGET_ARCH) -t debian)
 
+
+TARGET_ARCH_REDHAT = $(shell ./scripts/build/architecture-convert.sh -r $(TARGET_ARCH) -t redhat)
```
Comment (lurch) [inline Makefile]: ...and it probably makes sense for these TARGET_ARCH_... to be on consecutive lines, rather than having a blank line between them ;)

**Comment by lurch** [comment]
@Dark-Passenger Thanks for your work on this, it seems to be working great now :-)

Could you also please make the following changes, as that'll ensure that the RPM installers also get built by the CI servers. Don't try creating `bintray-redhat.sh` yourself though, that's something @jviotti needs to do as he's the only one with access to the bintray tokens.

```patch
diff --git a/Makefile b/Makefile
index eeb2382..967496f 100644
--- a/Makefile
+++ b/Makefile
@@ -476,6 +476,8 @@ PUBLISH_AWS_S3 += \
        $(BUILD_OUTPUT_DIRECTORY)/$(APPLICATION_NAME)-cli-$(APPLICATION_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).tar.gz
 PUBLISH_BINTRAY_DEBIAN += \
        $(BUILD_OUTPUT_DIRECTORY)/$(APPLICATION_NAME_LOWERCASE)-electron_$(APPLICATION_VERSION_DEBIAN)_$(TARGET_ARCH_DEBIAN).deb
+PUBLISH_BINTRAY_REDHAT += \
+       $(BUILD_OUTPUT_DIRECTORY)/$(APPLICATION_NAME_LOWERCASE)-electron_$(APPLICATION_VERSION_REDHAT)_$(TARGET_ARCH_REDHAT).rpm
 endif
 
 ifeq ($(TARGET_PLATFORM),win32)
@@ -492,7 +494,7 @@ PUBLISH_AWS_S3 += \
        $(BUILD_OUTPUT_DIRECTORY)/$(APPLICATION_NAME)-cli-$(APPLICATION_VERSION)-$(TARGET_PLATFORM)-$(TARGET_ARCH).zip
 endif
 
-installers-all: $(PUBLISH_AWS_S3) $(PUBLISH_BINTRAY_DEBIAN)
+installers-all: $(PUBLISH_AWS_S3) $(PUBLISH_BINTRAY_DEBIAN) $(PUBLISH_BINTRAY_REDHAT)
 
 ifdef PUBLISH_AWS_S3
 publish-aws-s3: $(PUBLISH_AWS_S3)
@@ -527,6 +529,13 @@ publish-bintray-debian: $(PUBLISH_BINTRAY_DEBIAN)
 TARGETS += publish-bintray-debian
 endif
 
+ifdef PUBLISH_BINTRAY_REDHAT
+publish-bintray-redhat: $(PUBLISH_BINTRAY_REDHAT)
+# TODO: Update this after we've created ./scripts/publish/bintray-redhat.sh
+
+TARGETS += publish-bintray-redhat
+endif
+
 .PHONY: $(TARGETS)
 
 cli-develop:
```

**Comment by jviotti** [comment]
Wow @Dark-Passenger, amazing job navigating our complex build system scripts 🥇 , and @lurch  amazing work reviewing and polishing this whole thing. Its the first time I check the PR code and its already incredibly good (I don't think I have any comments). I'll give it a go on my system.

File: scripts/build/redhat/config.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{
+  "name": "etcher-electron",
+  "productName": "Etcher",
+  "genericName": "Etcher",
+  "homepage": "https://etcher.io/",
+  "icon": "assets/icon.png",
```
Comment (jviotti) [inline scripts/build/redhat/config.json]: Maybe its just me, but the icon doesn't seem to be working on Fedora:

screenshot 2017-05-18 16 41 00

Comment (lurch) [inline scripts/build/redhat/config.json]: AFAICT this is the icon that gets displayed for the program e.g. in the applications menu?
I dunno enough about rpm to know if there's an additional icon that you can choose to display during installation? \*shrug\*  Maybe @Dark-Passenger knows... ;-)

**Comment by lurch** [comment]
@jviotti Did you read my comment about the 'missing' appveyor branch build?
This is what I see at the bottom of this PR:
image

File: scripts/build/redhat/config.json (hunk 5 above)
Comment (Dark-Passenger) [inline scripts/build/redhat/config.json]: That icon is the icon passed to the desktop application file(extension desktop) for the launcher icon. Gnome Software is supposed to use the same icon but I suppose its not getting the right size ? maybe we could have icons of other sizes ?

Comment (Dark-Passenger) [inline scripts/build/redhat/config.json]: Nope adding lower resolution icons doesnt help either. must be related to repository only download...

**Comment by lurch** [comment]
@jviotti Hmmm, seems like Appveyor is going crazy??

It failed with:
```
./scripts/build/electron-create-resources-app.sh -s . -o release/electron-win32-1.0.0+0c55bee-x86-app \
		-v 1.0.0+0c55bee \
		-f "lib,assets"
cp -RLf release/electron-win32-x86-dependencies/* release/electron-win32-1.0.0+0c55bee-x86-app
cp: cannot copy cyclic symbolic link `release/electron-win32-x86-dependencies/node_modules/@types/lodash/toString'
make: *** [release/electron-win32-1.0.0+0c55bee-x86-app] Error 1
Command exited with code 1
```
so I restarted the build, and now it's failing with:
```
> etcher@1.0.0 htmllint C:\projects\etcher
> node scripts/html-lint.js
Scanning...
2 errors at lib\gui\index.html
  [10:71] “--!>” found at end of comment (should just be “-->”).
  [11:69] “--!>” found at end of comment (should just be “-->”).
1 errors at lib\gui\pages\main\templates\main.tpl.html
  [66:93] “--!>” found at end of comment (should just be “-->”).
4 errors at lib\gui\pages\settings\templates\settings.tpl.html
  [20:67] “--!>” found at end of comment (should just be “-->”).
  [21:65] “--!>” found at end of comment (should just be “-->”).
  [22:66] “--!>” found at end of comment (should just be “-->”).
  [23:35] “--!>” found at end of comment (should just be “-->”).
```
but if I look at lines 10 and 11 of `lib\gui\index.html` they seem totally fine!?!
```
    <!-- Enable debug information from all modules that use `debug` -->
    <!-- See https://github.com/visionmedia/debug#browser-support -->
```

Gonna try another restart to see if the error changes yet again ;-)

BTW I also noticed this in the appveyor logs, which doesn't seem to be causing anything to fail:
```
./scripts/ci/ensure-staged-shrinkwrap.sh
node version: v6.10.3
npm version: 3.10.10
> etcher@1.0.0 preshrinkwrap C:\projects\etcher
> node ./scripts/clean-shrinkwrap.js
Removing: macos-alias, fs-xattr, ds-store, appdmg
events.js:160
      throw er; // Unhandled 'error' event
      ^
Error: spawn npm ENOENT
    at exports._errnoException (util.js:1018:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:193:32)
    at onErrorNT (internal/child_process.js:367:16)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
    at Module.runMain (module.js:606:11)
    at run (bootstrap_node.js:390:7)
    at startup (bootstrap_node.js:150:9)
    at bootstrap_node.js:505:3
wrote npm-shrinkwrap.json
```
Is it an error that's safe to ignore, or _should_ it be failing the build?

**Comment by lurch** [comment]
Ahhh, tracked this down to a bug in https://validator.w3.org (which is what `html-angular-validate` uses)
https://github.com/validator/validator/issues/515
:-/

Apologies for the noise @Dark-Passenger , please bear with us until we return to normal service ;)

**Comment by lurch** [comment]
Yaaay, https://github.com/validator/validator/issues/515 is now fixed, so I restarted the Appveyor build again, and now the pr-builds for Appveyor and Travis are both green :green_heart: 

Still the weird
```
./scripts/ci/ensure-staged-shrinkwrap.sh
node version: v6.10.3
npm version: 3.10.10
> etcher@1.0.0 preshrinkwrap C:\projects\etcher
> node ./scripts/clean-shrinkwrap.js
Removing: macos-alias, fs-xattr, ds-store, appdmg
events.js:160
      throw er; // Unhandled 'error' event
      ^
Error: spawn npm ENOENT
    at exports._errnoException (util.js:1018:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:193:32)
    at onErrorNT (internal/child_process.js:367:16)
    at _combinedTickCallback (internal/process/next_tick.js:80:11)
    at process._tickCallback (internal/process/next_tick.js:104:9)
    at Module.runMain (module.js:606:11)
    at run (bootstrap_node.js:390:7)
    at startup (bootstrap_node.js:150:9)
    at bootstrap_node.js:505:3
wrote npm-shrinkwrap.json
```
'error' on Appveyor though, and we still need to tell Appveyor not to branch-build this PR @jviotti

File: scripts/build/redhat/config.json (hunk 5 above)
Comment (jviotti) [inline scripts/build/redhat/config.json]: Yeah, looks like something we should configure somewhere else.

**Comment by jviotti** [comment]
@lurch I had indeed missed the Appveyor pending build thing. I'll revise the configuration. Regarding the `ENOENT` bug, I'm not sure what causes it, but it all seems to work fine anyway :S

**Comment by jviotti** [comment]
All green now! We were mandating Appveyor branch builds to succeed before merging, however in the case of PRs coming from forks, branch builds are not executed (Travis CI seems to handle this out of the box already)

**Comment by lurch** [comment]
Fantastic. I'll do some quick checks on Fedora x86 and Fedora x64 on Monday morning, and then merge this :-D

**Comment by Dark-Passenger** [comment]
@lurch @jviotti whats the status of this ? I was really hoping I could start using rpms soon for etcher :)

**Comment by jviotti** [comment]
@Dark-Passenger I'll merge as soon as @lurch gives the final approval

**Comment by lurch** [comment]
Sorry, I've just been incredibly busy with non-Etcher stuff recently. I'll try to get this merged in the next day or two, but if you could resolve the merge-conflicts @Dark-Passenger that would be helpful :-)

File: scripts/build/docker/Dockerfile-i686 [hunk 6]
```diff
@@ -29,8 +29,8 @@ RUN apt-get update \
     rsync \
     unzip \
     xvfb \
-    zip  \
-    rpm
+    xauth \
+    zip
```
Comment (lurch) [inline scripts/build/docker/Dockerfile-i686]: Looks like you accidentally removed `rpm` which is causing the Linux x86 Travis CI build to fail :-(

**Comment by Dark-Passenger** [comment]
@lurch anything left for this to be merged ?

**Comment by lurch** [comment]
Apologies again for the delay @Dark-Passenger , but @jviotti is now changing the way the Etcher build and packaging system works (with the aim of simplifying it and aligning it with other Electron tooling), so it's probably best if you hold off on this for a while, and we'll ping you again when the new system is all up and working.

**Comment by jviotti** [comment]
I don't have my PR ready yet, so I think we can merge this one, and I'll base my rpm work larlegy on what @Dark-Passenger did.

**Comment by lurch** [comment]
Okay, I'll do some tests in the next day or two and merge this as-is then :-)

**Comment by lurch** [comment]
In #1501 @jviotti says he's going to be changing the build system and re-testing everything anyway, so I'll merge this so that @jviotti can rebase his other PRs.

Thanks again @Dark-Passenger :+1:

**Pull request merged**